lib/teleterm/clusters/cluster_apps.go

```diff
@@ -36,3 +45,91 @@ type SAMLIdPServiceProvider struct {
 
 	Provider types.SAMLIdPServiceProvider
 }
+
+func (c *Cluster) getApp(ctx context.Context, appName string) (types.Application, error) {
+	var app types.Application
+	err := AddMetadataToRetryableError(ctx, func() error {
+		apps, err := c.clusterClient.ListApps(ctx, &proto.ListResourcesRequest{
+			Namespace:           c.clusterClient.Namespace,
+			ResourceType:        types.KindAppServer,
+			PredicateExpression: fmt.Sprintf(`name == "%s"`, appName),
+		})
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		if len(apps) == 0 {
+			return trace.NotFound("app %q not found", appName)
+		}
+
+		app = apps[0]
+		return nil
+	})
+
+	return app, trace.Wrap(err)
+}
+
+// reissueAppCert issue new certificates for the app and saves them to disk.
+func (c *Cluster) reissueAppCert(ctx context.Context, app types.Application) (tls.Certificate, error) {
+	if app.IsAWSConsole() || app.IsGCP() || app.IsAzureCloud() {
+		return tls.Certificate{}, trace.BadParameter("cloud applications are not supported")
+	}
+	// Refresh the certs to account for clusterClient.SiteName pointing at a leaf cluster.
+	err := c.clusterClient.ReissueUserCerts(ctx, client.CertCacheKeep, client.ReissueParams{
+		RouteToCluster: c.clusterClient.SiteName,
+		AccessRequests: c.status.ActiveRequests.AccessRequests,
+	})
+	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err)
+	}
+
+	proxyClient, err := c.clusterClient.ConnectToProxy(ctx)
+	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err)
+	}
+	defer proxyClient.Close()
+
+	request := types.CreateAppSessionRequest{
+		Username:          c.status.Username,
+		PublicAddr:        app.GetPublicAddr(),
+		ClusterName:       c.clusterClient.SiteName,
+		AWSRoleARN:        "",
+		AzureIdentity:     "",
+		GCPServiceAccount: "",
+	}
+
+	ws, err := proxyClient.CreateAppSession(ctx, request)
+	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err)
+	}
+
+	err = proxyClient.ReissueUserCerts(ctx, client.CertCacheKeep, client.ReissueParams{
+		RouteToCluster: c.clusterClient.SiteName,
+		RouteToApp: proto.RouteToApp{
+			Name:              app.GetName(),
+			SessionID:         ws.GetName(),
+			PublicAddr:        app.GetPublicAddr(),
+			ClusterName:       c.clusterClient.SiteName,
+			AWSRoleARN:        "",
+			AzureIdentity:     "",
+			GCPServiceAccount: "",
+		},
+		AccessRequests: c.status.ActiveRequests.AccessRequests,
+	})
+	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err)
+	}
+
+	key, err := c.clusterClient.LocalAgent().GetKey(c.clusterClient.SiteName, libclient.WithAppCerts{})
+	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err)
+	}
+
+	cert, ok := key.AppTLSCerts[app.GetName()]
+	if !ok {
+		return tls.Certificate{}, trace.NotFound("the user is not logged in into the application %v", app.GetName())
+	}
+
+	tlsCert, err := key.TLSCertificate(cert)
+	return tlsCert, trace.Wrap(err)
```

Comment:
FYI I took this from https://github.com/gravitational/teleport/blob/39af8d3268e281ed499a38524828378a42f91152/tool/tsh/common/proxy.go#L555